src/backend/x11/mod.rs

```diff
@@ -146,72 +126,124 @@ pub enum X11Event {
+/// Represents an active connection to the X.
+///
+/// Can be used to create a [`X11Backend`] or an [`smithay::backend::egl::EGLDisplay`].
 #[derive(Debug)]
-pub struct X11Backend {
-    log: Logger,
-    connection: Arc<RustConnection>,
-    source: X11Source,
-    inner: Arc<Mutex<X11Inner>>,
+pub struct X11Connection {
+    pub(crate) display: *mut xlib_sys::xlib::Display,
+    pub(crate) connection: Arc<XCBConnection>,
+    screen: usize,
+    logger: Logger,
 }
 
-impl X11Backend {
-    /// Initializes the X11 backend by connecting to the X server.
-    pub fn new<L>(logger: L) -> Result<X11Backend, X11Error>
+impl X11Connection {
+    /// Initializes the X11 connection to the X server.
+    pub fn new<L>(logger: L) -> Result<X11Connection, X11Error>
     where
         L: Into<Option<slog::Logger>>,
     {
         let logger = crate::slog_or_fallback(logger).new(o!("smithay_module" => "backend_x11"));
 
         info!(logger, "Connecting to the X server");
 
-        let (connection, screen_number) = RustConnection::connect(None)?;
+        let (connection, display, screen) = unsafe {
+            xlib_sys::xlib::XInitThreads();
+            let display = xlib_sys::xlib::XOpenDisplay(std::ptr::null());
+            if display.is_null() {
+                return Err(X11Error::ConnectionFailed(
+                    x11rb::errors::ConnectError::UnknownError,
+                ));
+            }
+            xlib_sys::xlib_xcb::XSetEventQueueOwner(
+                display,
+                xlib_sys::xlib_xcb::XEventQueueOwner::XCBOwnsEventQueue,
+            );
+            let screen = xlib_sys::xlib::XDefaultScreen(display);
+            let xcb_raw = xlib_sys::xlib_xcb::XGetXCBConnection(display);
+            (
+                XCBConnection::from_raw_xcb_connection(xcb_raw, true)?,
+                display,
+                screen as usize,
+            )
+        };
         let connection = Arc::new(connection);
-        info!(logger, "Connected to screen {}", screen_number);
+        info!(logger, "Connected to screen {}", screen);
+
+        let _ = Extensions::check_extensions(&*connection, &logger)?;
+
+        Ok(X11Connection {
+            display,
+            connection,
+            screen,
+            logger,
+        })
+    }
+}
+
+/// Represents an active connection to the X to manage events on the Window provided by the backend.
+#[derive(Debug)]
+pub struct X11Backend {
+    connection: Arc<XCBConnection>,
+    source: X11Source<XCBConnection>,
+    inner: Arc<Mutex<X11Inner>>,
+    log: Logger,
+}
+
+unsafe impl Send for X11Connection {}
+unsafe impl Sync for X11Connection {}
 
-        let extensions = Extensions::check_extensions(&*connection, &logger)?;
+impl X11Backend {
+    /// Initializes the X11 backend by connecting to the X server.
+    pub fn new(connection: X11Connection, context: &EGLContext) -> Result<X11Backend, X11Error> {
```

Comment:
This is pribably too early to involve egl, I'd prefer egl getting involved at the point of window creation or as a sort of X11EglSurface